genkit-tools/common/src/eval/localFileDatasetStore.ts

```diff
@@ -201,8 +201,34 @@ export class LocalFileDatasetStore implements DatasetStore {
     return path.resolve(process.cwd(), `.genkit/${rootHash}/datasets`);
   }
 
-  private generateDatasetId(): string {
-    return uuidv4();
+  /** Visible for testing */
+  async generateDatasetId(
+    displayName?: string,
+    datasetId?: string
+  ): Promise<string> {
+    if (datasetId) {
+      return datasetId;
+    }
+
+    const metadataMap = await this.getMetadataMap();
+    const keys = Object.keys(metadataMap);
+    const trimmedDisplayName =
+      displayName?.replaceAll(/[^A-Za-z0-9-_]/g, '') ?? '';
+    const uuidLength = displayName ? this.SHORT_UUID_LENGTH : undefined;
+
+    for (let i = 0; i < this.ID_GEN_MAX_ATTEMPTS; i++) {
+      const uid = uuidv4().slice(0, uuidLength);
+      const maybeDatasetId = trimmedDisplayName
+        ? `${trimmedDisplayName}-${uid}`
```

Comment:
As discussed offline, lets either truncate the display name too, to enforce some max length on the ID or set a max length on display name.